x-pack/plugins/reporting/server/export_types/png_v2/execute_job.ts

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import apm from 'elastic-apm-node';
+import * as Rx from 'rxjs';
+import { catchError, finalize, map, mergeMap, takeUntil } from 'rxjs/operators';
+import { PNG_JOB_TYPE_V2, getRedirectAppPathHome } from '../../../common/constants';
+import { TaskRunResult } from '../../lib/tasks';
+import { RunTaskFn, RunTaskFnFactory } from '../../types';
+import { decryptJobHeaders, getConditionalHeaders, omitBlockedHeaders } from '../common';
+import { getFullUrls } from '../common/v2/get_full_urls';
+import { generatePngObservableFactory } from './lib/generate_png';
+import { TaskPayloadPNGV2 } from './types';
+
+export const runTaskFnFactory: RunTaskFnFactory<
+  RunTaskFn<TaskPayloadPNGV2>
+> = function executeJobFactoryFn(reporting, parentLogger) {
+  const config = reporting.getConfig();
+  const encryptionKey = config.get('encryptionKey');
+
+  return async function runTask(jobId, job, cancellationToken) {
+    const apmTrans = apm.startTransaction('reporting execute_job pngV2', 'reporting');
+    const apmGetAssets = apmTrans?.startSpan('get_assets', 'setup');
+    let apmGeneratePng: { end: () => void } | null | undefined;
+
+    const generatePngObservable = await generatePngObservableFactory(reporting);
+    const jobLogger = parentLogger.clone([PNG_JOB_TYPE_V2, 'execute', jobId]);
+    const process$: Rx.Observable<TaskRunResult> = Rx.of(1).pipe(
+      mergeMap(() => decryptJobHeaders(encryptionKey, job.headers, jobLogger)),
+      map((decryptedHeaders) => omitBlockedHeaders(decryptedHeaders)),
+      map((filteredHeaders) => getConditionalHeaders(config, filteredHeaders)),
+      mergeMap((conditionalHeaders) => {
+        const { locatorParams } = job;
+        const relativeUrls = locatorParams.map((_, idx) =>
```

Comment:
I updated the `createJob` interface to expect a single `locatorParams` but internally I still store as an array. From the user perspective this will only ever allow one locator, but internally we store an array with one item - this seemed like a nice way to keep things simpler from a code perspective. Do you have concerns about that approach?

Reply:
wouldn't code be simpler if we just had one item instead of array of items internally as well ?

Reply:
For PDFs we possibly have multiple. I was thinking it would simplify the code to store them all as an array (even for PNGs). Unless I'm misunderstanding your suggestion - are you thinking we create a saved object per page?